framework/src/mfem/equation_systems/EquationSystem.C

```diff
@@ -476,50 +496,57 @@ TimeDependentEquationSystem::SetTimeStep(mfem::real_t dt)
 void
 TimeDependentEquationSystem::SetTrialVariableNames()
 {
-  // If a coupled variable has an equation associated with it,
-  // add it to the set of trial variables.
-  for (const auto & coupled_var_name : _coupled_var_names)
+  // The TimeDependentEquationSystem operator expects to act on a vector of variable time
+  // derivatives, so the trial variable must be the time derivative of the 'base' variable. The base
+  // variable (test_var_name) without derivatives applied must also be coupled in implicit
+  // timestepping schemes for the elimination of 'old' variable values from the previous timestep
+  for (const auto & test_var_name : _test_var_names)
   {
-    for (const auto & test_var_name : _test_var_names)
+    AddCoupledVariableNameIfMissing(test_var_name);
+    AddCoupledVariableNameIfMissing(_time_derivative_map.getTimeDerivativeName(test_var_name));
+  }
+
+  // If a coupled variable does not have an equation associated with it,
+  // add it to the set of eliminated variables.
+  for (const auto & test_var_name : _test_var_names)
+  {
+    const auto time_derivative_test_var_name =
+        _time_derivative_map.getTimeDerivativeName(test_var_name);
+    for (const auto & coupled_var_name : _coupled_var_names)
     {
-      const auto time_derivative_test_var_name = GetTimeDerivativeName(test_var_name);
-      if (time_derivative_test_var_name == coupled_var_name)
-      {
-        if (!VectorContainsName(_trial_var_names, coupled_var_name))
-          _trial_var_names.push_back(coupled_var_name);
-      }
-      else
-      {
-        if (!VectorContainsName(_eliminated_var_names, coupled_var_name))
-          _eliminated_var_names.push_back(coupled_var_name);
-      }
+      if (time_derivative_test_var_name != coupled_var_name &&
+          !VectorContainsName(_eliminated_var_names, coupled_var_name))
+        _eliminated_var_names.push_back(coupled_var_name);
     }
+    if (!VectorContainsName(_trial_var_names, time_derivative_test_var_name))
+      _trial_var_names.push_back(time_derivative_test_var_name);
   }
 }
 
 void
 TimeDependentEquationSystem::AddKernel(std::shared_ptr<MFEMKernel> kernel)
 {
-  if (kernel->getTrialVariableName() == GetTimeDerivativeName(kernel->getTestVariableName()))
+  auto trial_var_name = kernel->getTrialVariableName();
+  auto test_var_name = kernel->getTestVariableName();
+
+  if (_time_derivative_map.isTimeDerivative(trial_var_name))
   {
-    auto trial_var_name = kernel->getTrialVariableName();
-    auto test_var_name = kernel->getTestVariableName();
     AddTestVariableNameIfMissing(test_var_name);
-    AddCoupledVariableNameIfMissing(test_var_name);
+    AddCoupledVariableNameIfMissing(trial_var_name);
```

Comment:
The mix of lower camel case and upper camel case still dizzies me a bit. I understand that this is inheriting from `mfem::Operator`. Is that why the method names for this class are upper camel case? Do we document the convention anywhere?

Reply:
Yes, that was the original intent, with the logic being that the upper camel case convention was inherited from `mfem::Operator` in order to ensure newly added methods shared the same case convention as methods inherited from the base class. The same applies to the `ProblemOperator` classes for the same reason